numpy/_core/src/multiarray/item_selection.c

```diff
@@ -2797,6 +2797,13 @@ NPY_NO_EXPORT PyObject *
 PyArray_Nonzero(PyArrayObject *self)
 {
     int i, ndim = PyArray_NDIM(self);
+    if (ndim == 0) {
+        PyErr_SetString(PyExc_ValueError,
+            "Calling nonzero on 0d arrays is not allowed. "
+            "Use np.atleast_1d(scalar).nonzero() instead.");
```

Comment:
Can you leave the slightly nicer error message for bool arrays from the old deprecated warning?

Reply:
Done!